File: lib/src/stitching/stitching.dart
// ignore_for_file: constant_identifier_names

library cv.stitching;

import 'dart:ffi' as ffi;

import 'package:ffi/ffi.dart';

import '../core/base.dart';
import '../core/mat.dart';
import '../core/vec.dart';
import '../g/stitching.g.dart' as cvg;
import '../native_lib.dart' show cstitching;

/// High level image stitcher.
///
/// It's possible to use this class without being aware of the entire
/// stitching pipeline. However, to be able to achieve higher stitching
/// stability and quality of the final images at least being familiar
/// with the theory is recommended.
/// https://docs.opencv.org/4.x/d2/d8d/classcv_1_1Stitcher.html#details
class Stitcher extends CvStruct<cvg.Stitcher> {
  Stitcher._(cvg.StitcherPtr ptr, [bool attach = true]) : super.fromPointer(ptr) {
    if (attach) {
      finalizer.attach(this, ptr.cast(), detach: this);
    }
  }
  factory Stitcher.fromPointer(cvg.StitcherPtr ptr, [bool attach = true]) => Stitcher._(ptr.cast(), attach);

  /// Creates a Stitcher configured in one of the stitching modes.
  /// https://docs.opencv.org/4.x/d2/d8d/classcv_1_1Stitcher.html#a308a47865a1f381e4429c8ec5e99549f
  factory Stitcher.create({StitcherMode mode = StitcherMode.PANORAMA}) {
    final ptr_ = calloc<cvg.Stitcher>();
    cvRun(() => cstitching.Stitcher_Create(mode.index, ptr_));
    return Stitcher._(ptr_);
  }

  /// https://docs.opencv.org/4.x/d2/d8d/classcv_1_1Stitcher.html#a9b90774eabdf68c9ee864918d620538d
  double get registrationResol {
    return using<double>((arena) {
      final rptr = arena<ffi.Double>();
      cvRun(() => cstitching.Stitcher_GetRegistrationResol(ref, rptr));
      return rptr.value;
    });
  }

  /// https://docs.opencv.org/4.x/d2/d8d/classcv_1_1Stitcher.html#a9912fe8c095b8385267908e5ef707439
  set registrationResol(double value) {
    cvRun(() => cstitching.Stitcher_SetRegistrationResol(ref, value));
  }

  /// https://docs.opencv.org/4.x/d2/d8d/classcv_1_1Stitcher.html#ac559c3eb228614f9402ff3eba23a08f5
  double get seamEstimationResol {
    return using<double>((arena) {
      final rptr = arena<ffi.Double>();
      cvRun(() => cstitching.Stitcher_GetSeamEstimationResol(ref, rptr));
      return rptr.value;
    });
  }

  /// https://docs.opencv.org/4.x/d2/d8d/classcv_1_1Stitcher.html#ad0fcef52b2fedda1dbb90ea780cd7979
  set seamEstimationResol(double value) {
    cvRun(() => cstitching.Stitcher_SetSeamEstimationResol(ref, value));
  }

  /// https://docs.opencv.org/4.x/d2/d8d/classcv_1_1Stitcher.html#ad13d2d50b253e471fbaf041b9a044571
  double get compositingResol {
    return using<double>((arena) {
      final rptr = arena<ffi.Double>();
      cvRun(() => cstitching.Stitcher_GetCompositingResol(ref, rptr));
      return rptr.value;
    });
  }

  /// https://docs.opencv.org/4.x/d2/d8d/classcv_1_1Stitcher.html#afe927e80fcb2ca2061630ddd98eebba8
  set compositingResol(double value) {
    cvRun(() => cstitching.Stitcher_SetCompositingResol(ref, value));
  }

  /// https://docs.opencv.org/4.x/d2/d8d/classcv_1_1Stitcher.html#a3755bbeca7f4c80dc42af034f7621568
  double get panoConfidenceThresh {
    return using<double>((arena) {
      final rptr = arena<ffi.Double>();
      cvRun(() => cstitching.Stitcher_GetPanoConfidenceThresh(ref, rptr));
      return rptr.value;
    });
  }

  /// https://docs.opencv.org/4.x/d2/d8d/classcv_1_1Stitcher.html#a6f5e62bc1dd5d7bdb5f9313a2c21c558
  set panoConfidenceThresh(double value) {
    cvRun(() => cstitching.Stitcher_SetPanoConfidenceThresh(ref, value));
  }

  /// https://docs.opencv.org/4.x/d2/d8d/classcv_1_1Stitcher.html#af6a51e0b23dac119a3612d57345f9a7f
  bool get waveCorrection {
    return using<bool>((arena) {
      final rptr = arena<ffi.Bool>();
      cvRun(() => cstitching.Stitcher_GetWaveCorrection(ref, rptr));
      return rptr.value;
    });
  }

  /// https://docs.opencv.org/4.x/d2/d8d/classcv_1_1Stitcher.html#a968a2f4a1faddfdacbcfce54b44bab70
  set waveCorrection(bool value) {
    cvRun(() => cstitching.Stitcher_SetWaveCorrection(ref, value));
  }

  /// https://docs.opencv.org/4.x/d2/d8d/classcv_1_1Stitcher.html#abc0c8f54a1d223a1098206654813d973
  int get interpolationFlags {
    return using<int>((arena) {
      final rptr = arena<ffi.Int>();
      cvRun(() => cstitching.Stitcher_GetInterpolationFlags(ref, rptr));
      return rptr.value;
    });
  }

  /// https://docs.opencv.org/4.x/d2/d8d/classcv_1_1Stitcher.html#a253d04b8dcd3c674321b29139c769873
  set interpolationFlags(int value) {
    cvRun(() => cstitching.Stitcher_SetInterpolationFlags(ref, value));
  }

  /// https://docs.opencv.org/4.x/d2/d8d/classcv_1_1Stitcher.html#ad9c9c9b8a97b686ad3b93f7918c4c6de
  int get waveCorrectKind {
    return using<int>((arena) {
      final rptr = arena<ffi.Int>();
      cvRun(() => cstitching.Stitcher_GetWaveCorrectKind(ref, rptr));
      return rptr.value;
    });
  }

  /// https://docs.opencv.org/4.x/d2/d8d/classcv_1_1Stitcher.html#a17413f5c06e4e569bfd45e01d4e8ff4a
  set waveCorrectKind(int value) {
    cvRun(() => cstitching.Stitcher_SetWaveCorrectKind(ref, value));
  }

  /// These functions try to match the given images and to estimate rotations of each camera.
  /// https://docs.opencv.org/4.x/d2/d8d/classcv_1_1Stitcher.html#a4c25557af4d40a79a4d1f23d9548131d
  StitcherStatus estimateTransform(VecMat images, {VecMat? masks}) {
    return using<StitcherStatus>((arena) {
      final rptr = arena<ffi.Int>();
      masks ??= VecMat.fromList([]);
      cvRun(() => cstitching.Stitcher_EstimateTransform(ref, images.ref, masks!.ref, rptr));
      return StitcherStatus.fromInt(rptr.value);
    });
  }

  /// These functions try to compose the given images (or images stored internally
  /// from the other function calls) into the final pano under the assumption
  /// that the image transformations were estimated before.
  /// https://docs.opencv.org/4.x/d2/d8d/classcv_1_1Stitcher.html#acc8409a6b2e548de1653f0dc5c2ccb02
  (StitcherStatus, Mat pano) composePanorama({VecMat? images}) {
    return using<(StitcherStatus, Mat)>((arena) {
      final rptr = arena<ffi.Int>();
      final rpano = Mat.empty();
      images == null
          ? cvRun(() => cstitching.Stitcher_ComposePanorama(ref, rpano.ref, rptr))
          : cvRun(() => cstitching.Stitcher_ComposePanorama_1(ref, images.ref, rpano.ref, rptr));
      return (StitcherStatus.fromInt(rptr.value), rpano);
    });
  }

  /// This is an overloaded member function, provided for convenience.
  /// It differs from the above function only in what argument(s) it accepts.
  /// https://docs.opencv.org/4.x/d2/d8d/classcv_1_1Stitcher.html#a37ee5bacf229e9d0fb9f97c8f5ed1acd
  (StitcherStatus, Mat pano) stitch(VecMat images, {VecMat? masks}) {
    return using<(StitcherStatus, Mat)>((arena) {
      final rptr = arena<ffi.Int>();
      final rpano = Mat.empty();
      masks == null
          ? cvRun(() => cstitching.Stitcher_Stitch(ref, images.ref, rpano.ref, rptr))
          : cvRun(() => cstitching.Stitcher_Stitch_1(ref, images.ref, masks.ref, rpano.ref, rptr));
      return (StitcherStatus.fromInt(rptr.value), rpano);
    });
  }

  /// https://docs.opencv.org/4.x/d2/d8d/classcv_1_1Stitcher.html#a7fed80561a9b46a1a924ac6cb334ac85
  VecI32 get component {
    final rptr = calloc<cvg.VecI32>();
    cvRun(() => cstitching.Stitcher_Component(ref, rptr));
    return VecI32.fromPointer(rptr);
  }

  static final finalizer = OcvFinalizer<cvg.StitcherPtr>(cstitching.addresses.Stitcher_Close);

  void dispose() {
    finalizer.detach(this);
    cstitching.Stitcher_Close(ptr);
  }

  @override
  cvg.Stitcher get ref => ptr.ref;
}

/// https://docs.opencv.org/4.x/d2/d8d/classcv_1_1Stitcher.html#a507409ce9435dd89857469d12ec06b45
enum StitcherStatus {
  OK,
  ERR_NEED_MORE_IMGS,
  ERR_HOMOGRAPHY_EST_FAIL,
  ERR_CAMERA_PARAMS_ADJUST_FAIL;

  factory StitcherStatus.fromInt(int v) => values.firstWhere((e) => e.index == v);
}

/// https://docs.opencv.org/4.x/d2/d8d/classcv_1_1Stitcher.html#a114713924ec05a0309f4df7e918c0324
enum StitcherMode {
  /// Mode for creating photo panoramas. Expects images under perspective transformation and projects resulting pano to sphere.
  PANORAMA,

  /// Mode for composing scans. Expects images under affine transformation does not compensate exposure by default.
  SCANS;

  factory StitcherMode.fromInt(int v) => values.firstWhere((e) => e.index == v);
}

/// https://docs.opencv.org/4.x/d7/d74/group__stitching__rotation.html#ga83b24d4c3e93584986a56d9e43b9cf7f
enum WaveCorrectKind { HORIZONTAL, VERTICAL }

File: src/stitching/stitching.cpp
/*
    Created by Rainyl.
    Licensed: Apache 2.0 license. Copyright (c) 2024 Rainyl.
*/

#include "stitching.h"
#include "core/vec.hpp"

CvStatus *Stitcher_Create(int mode, Stitcher *rval) {
  BEGIN_WRAP
  const auto ptr = cv::Stitcher::create(static_cast<cv::Stitcher::Mode>(mode));
  *rval = {new cv::Ptr<cv::Stitcher>(ptr)};
  END_WRAP
}

void Stitcher_Close(StitcherPtr stitcher) {
  stitcher->ptr->reset();
  CVD_FREE(stitcher);
}

CvStatus *Stitcher_GetRegistrationResol(Stitcher self, double *rval) {
  BEGIN_WRAP
  *rval = (*self.ptr)->registrationResol();
  END_WRAP
}

CvStatus *Stitcher_SetRegistrationResol(Stitcher self, double inval) {
  BEGIN_WRAP(*self.ptr)->setRegistrationResol(inval);
  END_WRAP
}

CvStatus *Stitcher_GetSeamEstimationResol(Stitcher self, double *rval) {
  BEGIN_WRAP
  *rval = (*self.ptr)->seamEstimationResol();
  END_WRAP
}
CvStatus *Stitcher_SetSeamEstimationResol(Stitcher self, double inval) {
  BEGIN_WRAP(*self.ptr)->setSeamEstimationResol(inval);
  END_WRAP
}

CvStatus *Stitcher_GetCompositingResol(Stitcher self, double *rval) {
  BEGIN_WRAP
  *rval = (*self.ptr)->compositingResol();
  END_WRAP
}
CvStatus *Stitcher_SetCompositingResol(Stitcher self, double inval) {
  BEGIN_WRAP(*self.ptr)->setCompositingResol(inval);
  END_WRAP
}

CvStatus *Stitcher_GetPanoConfidenceThresh(Stitcher self, double *rval) {
  BEGIN_WRAP
  *rval = (*self.ptr)->panoConfidenceThresh();
  END_WRAP
}
CvStatus *Stitcher_SetPanoConfidenceThresh(Stitcher self, double inval) {
  BEGIN_WRAP(*self.ptr)->setPanoConfidenceThresh(inval);
  END_WRAP
}

CvStatus *Stitcher_GetWaveCorrection(Stitcher self, bool *rval) {
  BEGIN_WRAP
  *rval = (*self.ptr)->waveCorrection();
  END_WRAP
}
CvStatus *Stitcher_SetWaveCorrection(Stitcher self, bool inval) {
  BEGIN_WRAP(*self.ptr)->setWaveCorrection(inval);
  END_WRAP
}

CvStatus *Stitcher_GetInterpolationFlags(Stitcher self, int *rval) {
  BEGIN_WRAP
  *rval = static_cast<int>((*self.ptr)->interpolationFlags());
  END_WRAP
}
CvStatus *Stitcher_SetInterpolationFlags(Stitcher self, int inval) {
  BEGIN_WRAP(*self.ptr)->setInterpolationFlags(static_cast<cv::InterpolationFlags>(inval));
  END_WRAP
}

CvStatus *Stitcher_GetWaveCorrectKind(Stitcher self, int *rval) {
  BEGIN_WRAP
  *rval = static_cast<int>((*self.ptr)->waveCorrectKind());
  END_WRAP
}
CvStatus *Stitcher_SetWaveCorrectKind(Stitcher self, int inval) {
  BEGIN_WRAP(*self.ptr)->setWaveCorrectKind(static_cast<cv::detail::WaveCorrectKind>(inval));
  END_WRAP
}

CvStatus *Stitcher_EstimateTransform(Stitcher self, VecMat mats, VecMat masks, int *rval) {
  BEGIN_WRAP
  auto _mats = vecmat_c2cpp(mats);
  if (masks.length > 0) {
    auto _masks = vecmat_c2cpp(masks);
    *rval = static_cast<int>((*self.ptr)->estimateTransform(_mats, _masks));
  } else
    *rval = static_cast<int>((*self.ptr)->estimateTransform(_mats));
  END_WRAP
}

CvStatus *Stitcher_ComposePanorama(Stitcher self, Mat rpano, int *rval) {
  BEGIN_WRAP
  *rval = static_cast<int>((*self.ptr)->composePanorama(*rpano.ptr));
  END_WRAP
}
CvStatus *Stitcher_ComposePanorama_1(Stitcher self, VecMat mats, Mat rpano, int *rval) {
  BEGIN_WRAP
  auto _mats = vecmat_c2cpp(mats);
  *rval = static_cast<int>((*self.ptr)->composePanorama(_mats, *rpano.ptr));
  END_WRAP
}

CvStatus *Stitcher_Stitch(Stitcher self, VecMat mats, Mat rpano, int *rval) {
  BEGIN_WRAP
  auto _mats = vecmat_c2cpp(mats);
  *rval = static_cast<int>((*self.ptr)->stitch(_mats, *rpano.ptr));
  END_WRAP
}
CvStatus *Stitcher_Stitch_1(Stitcher self, VecMat mats, VecMat masks, Mat rpano, int *rval) {
  BEGIN_WRAP
  auto _mats = vecmat_c2cpp(mats);
  auto _masks = vecmat_c2cpp(masks);
  *rval = static_cast<int>((*self.ptr)->stitch(_mats, _masks, *rpano.ptr));
  END_WRAP
}

CvStatus *Stitcher_Component(Stitcher self, VecI32 *rval) {
  BEGIN_WRAP
  std::vector<int> _rval = (*self.ptr)->component();
  *rval = vecint_cpp2c(_rval);
  END_WRAP
}
